fix: backward compatibility with version 8.0.0

salvatore-campagna opened the pull request:
Commit `fb6e1fb127e3c2bb1683402e9ab083115a56b08d` did not land into version `8.0.0`. As a result, we need to make master (version `8.2.0`) interoperate with version `8.0.0`.
This means we need to do the following:
* always read and write a **non-optional string key** for a range bucket (not for the range itself but for the bucket). In that case the key is used when serialising the response back to clients.
* write `originalFrom` and `originalTo` values so that a node running version `8.0.0` can read them properly.
* read `originalFrom` and `originalTo` from a node running version `8.0.0` and using them to **overwrite** `from` and `to` when creating a new range bucket.

All of this needs to be done keeping compatibility with version `7.17` which was fixed by PR #83715 .
Also note that, interoperability with version `8.0.1` is taken care by PR #83715. As a result, here we really just need to address backward compatibility with version `8.0.0` that is already released and we cannot change.

salvatore-campagna committed:
fix: stop reading and writing two optional and unused double values

With commit 3550370c86cfdee60bb5ca991805437f0b87bccc two values where
introduced (originalFrom and originalTo) which have been made redundant
later by commit a03cd922079f55961b28afa8c8a3766924890850.

As a result we stop writing and reading unused values from version
8.2.0.

elasticmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

salvatore-campagna commented:
@elasticmachine run elasticsearch-ci/part-1

salvatore-campagna commented:
@elasticmachine run elasticsearch-ci/part-2

salvatore-campagna commented:
@elasticmachine run elasticsearch-ci/packaging-tests-unix-sample

salvatore-campagna commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/range/InternalRange.java:
```diff
@@ -265,11 +265,11 @@ public InternalRange(StreamInput in) throws IOException {
             double from = in.readDouble();
-            if (in.getVersion().onOrAfter(Version.V_7_17_0)) {
+            if (in.getVersion().onOrAfter(Version.V_7_17_0) && in.getVersion().before(Version.V_8_2_0)) {
                 in.readOptionalDouble();
```
As a result of #83175 here we need to overwrite both `from` and `to` with whatever value we read from the wire if version matches `7.17.0`. Version `7.17.0` indeed, writes `from`, `originalFrom`, `to`, `originalTo` on the wire. In that case, if we just drop `originalFrom` and `originalTo` we have incorrect results because we would use the `from` and `to` during serialisation, which have precision issues.

nik9000 reviewed (approved):
LGTM though we should wait until #83715 is in and settled.

salvatore-campagna commented:
@elasticmachine update branch

elasticmachine commented:
merge conflict between base and head

salvatore-campagna committed:
fix: handle possible null value for optional double

salvatore-campagna committed:
Merge branch 'master' into fix/stop-reading-unused-values

salvatore-campagna committed:
fix: backward compatibility with 8.0.0

Version 8.0.0 always expects to read and write a non-null key for
the range bucket.

salvatore-campagna committed:
fix: code format violations

nik9000 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/range/InternalBinaryRange.java:
```diff
@@ -70,7 +70,9 @@ private static String generateKey(BytesRef from, BytesRef to, DocValueFormat for
 
         private static Bucket createFromStream(StreamInput in, DocValueFormat format, boolean keyed) throws IOException {
-            String key = in.getVersion().onOrAfter(Version.V_7_17_1) ? in.readOptionalString() : in.readString();
+            String key = in.getVersion().equals(Version.V_8_0_0) ? in.readString()
```
Maybe good to put a comment here in word saying what this does.

nik9000 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/range/InternalBinaryRange.java:
```diff
@@ -81,10 +83,14 @@ private static Bucket createFromStream(StreamInput in, DocValueFormat format, bo
-            } else {
+            if (out.getVersion().equals(Version.V_8_0_0)) {
                 out.writeString(key == null ? generateKey(from, to, format) : key);
+            } else {
```
Maybe more readable with `} else if {`

nik9000 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/range/InternalRange.java:
```diff
@@ -157,17 +157,21 @@ private static String generateKey(double from, double to, DocValueFormat format)
             }
             out.writeDouble(from);
-            if (out.getVersion().onOrAfter(Version.V_7_17_0)) {
+            if (out.getVersion().onOrAfter(Version.V_7_17_0) && out.getVersion().before(Version.V_8_2_0)) {
```
I think you should move the v_8_2_0 thing to a different PR. it'll save a byte which is great but it makes the PR hard to backport.

salvatore-campagna committed:
fix: address review comments

salvatore-campagna committed:
fix: use else if instead of elase with nested if

salvatore-campagna committed:
fix: add a note explaining bucket key requirements

salvatore-campagna committed:
fix: note wording mistake

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.1](https://github.com/elastic/elasticsearch/pull/83908)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/83908">](https://github.com/elastic/elasticsearch/pull/83908) |
| ✅ |  [8.0](https://github.com/elastic/elasticsearch/pull/83909)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/83909">](https://github.com/elastic/elasticsearch/pull/83909) |
